Eth account support (#361)
OpenZeppelin/cairo-contracts: docs/Account.md

  * [`set_public_key`](#set_public_key)
  * [`is_valid_signature`](#is_valid_signature)
  * [`__execute__`](#__execute__)
  * [`is_valid_eth_signature`](#is_valid_eth_signature)
  * [`eth_execute`](#eth_execute)
  * [`_unsafe_execute`](#_unsafe_execute)
* [Presets](#presets)

>   * [Eth Account](#eth-account)

Reviewer: Good call adding these in! We should add headers for `Presets` and `Eth Account` in the docs (the links don't point to anything right now). I think our vanilla `Account` should be in `Presets` as well

This is the only external entrypoint to interact with the Account contract. It:

1. Validates the transaction signature matches the message (including the nonce)

> 3. Increments the nonce
> 4. Calls the target contract with the intended function selector and calldata parameters
> 5. Forwards the contract call response data as return value

Reviewer 2: ```suggestion
2. Increments the nonce
3. Calls the target contract with the intended function selector and calldata parameters
4. Forwards the contract call response data as return value
```

This is the only external entrypoint to interact with the Account contract. It:

1. Validates the transaction signature matches the message (including the nonce)

> 2. Calls _unsafe_execute.

Reviewer 2: `__execute__` doesn't call `_unsafe_execute` directly! That's a subcall implementation detail. This section is confusing the external preset function with the library API. We need to distinguish between preset and library APIs -- something we've mentioned needs improvement across for the docs. For the time being, let's at least document in here that each preset's `__execute__` will call its corresponding library counterpart.

Then on each library function API doc (`execute` and `eth_execute`), we document they call `_unsafe_execute`.

response: felt*
```

### `_unsafe_execute`

> Is the internal method that performs the following task inside the `execute` and `eth_execute` functions:

Reviewer 2: ```suggestion
It's an internal method that performs the following tasks:
```
`_unsafe_execute` is not aware of who's calling it, it may be used by other `execute` flavors too.

1. Increments the nonce.
2. Takes the input and builds a `Call` for each iterated message. See [Multicall transactions](#multicall-transactions) for more information.
3. Calls the target contract with the intended function selector and calldata parameters
4. Forwards the contract call response data as return value

## Presets

The following contract presets are ready to deploy and can be used as-is for quick prototyping and testing. Each preset differs on the signature type being used by the Account.

### Account

> The [`Account`](../src/openzeppelin/account/Account.cairo) preset offers an basic implementation of the library.

Reviewer 2: ```suggestion
The [`Account`](../src/openzeppelin/account/Account.cairo) preset uses StarkNet keys to validate transactions.
```

### Eth Account

> The [`EthAccount`](../src/openzeppelin/account/EthAccount.cairo) preset offers an account that supports ethereum addresses, meaning the key pair uses the 256k1 curve.

Reviewer 2: ```suggestion
The [`EthAccount`](../src/openzeppelin/account/EthAccount.cairo) preset supports Ethereum addresses, validating transactions with secp256k1 keys.
```

> Account contracts can be extended by following the [extensibility pattern](../docs/Extensibility.md#the-pattern). The basic idea behind integrating the pattern is to import the requisite methods from the Account library and incorporate the extended logic thereafter.

Reviewer 2: ```suggestion
Account contracts can be extended by following the [extensibility pattern](../docs/Extensibility.md#the-pattern).
```

> Currently, there's a library/preset Account scheme, and an EthAccount preset, both contain a validation and an execution step, where `_unsafe_execution` is used as a building block for each scheme, and exposing it is highly discourage, but we're looking for feedback and new presets to emerge. Some new validation schemes to look out for in the future:

Reviewer 2: ```suggestion
To implement custom account contracts, a pair of `validate` and `execute` functions should be exposed. This is why the Account library comes with different flavors of such pairs, like the vanilla `is_valid_signature` and `execute`, or the Ethereum flavored `is_valid_eth_signature` and `eth_execute` pair.

Account contract developers are encouraged to implement the [standard Account interface](https://github.com/OpenZeppelin/cairo-contracts/discussions/41) and incorporate the custom logic thereafter.

To implement alternative `execute` functions, make sure to check their corresponding `validate` function before calling the `_unsafe_execution` building block. Do not expose `_unsafe_execution` directly.

Some other validation schemes to look out for in the future:
```

* multisig
* guardian logic like in [Argent's account](https://github.com/argentlabs/argent-contracts-starknet/blob/de5654555309fa76160ba3d7393d32d2b12e7349/contracts/ArgentAccount.cairo)

> * [Ethereum signatures](https://github.com/OpenZeppelin/cairo-contracts/issues/161)

Reviewer: Seconded^. Maybe add an underscore to emphasize its discouraged use as-is? => `_unsafe_execution`

> * [Ethereum signatures](https://github.com/OpenZeppelin/cairo-contracts/issues/161)

Reviewer 2: but it's part of the exported api for building custom accounts, isn't it?
